src/persist-types/src/part.rs

```diff
@@ -223,72 +224,99 @@ impl Part {
 /// An in-progress columnar constructor for one blob's worth of data.
 #[derive(Debug)]
 pub struct PartBuilder {
-    key: DynStructMut,
-    val: DynStructMut,
-    ts: Vec<i64>,
-    diff: Vec<i64>,
+    /// The key column.
+    pub key: ColumnsMut,
+    /// The val column.
+    pub val: ColumnsMut,
+    /// The ts column.
+    pub ts: Codec64Mut,
+    /// The diff column.
+    pub diff: Codec64Mut,
 }
 
 impl PartBuilder {
-    /// Returns a new PartBuilder with the given schema.
-    pub fn new<K, KS: Schema<K>, V, VS: Schema<V>>(key_schema: &KS, val_schema: &VS) -> Self {
-        let key = ColumnMut::<DynStructCfg>::new(&key_schema.columns());
-        let val = ColumnMut::<DynStructCfg>::new(&val_schema.columns());
-        let ts = Vec::new();
-        let diff = Vec::new();
-        PartBuilder { key, val, ts, diff }
-    }
+    /// Returns a new [`PartBuilder`] and [`PartBuilderCfg`] with the given schema.
+    ///
+    /// See [`PartBuilderCfg::into_part`] for how to finalize into a [`Part`].
+    pub fn new<K, KS: Schema<K>, V, VS: Schema<V>>(
+        key_schema: &KS,
+        val_schema: &VS,
+    ) -> (PartBuilderCfg, Self) {
```

Comment:
> From: @danhhz in https://github.com/MaterializeInc/materialize/pull/26605#discussion_r1567880446
> Naive question: I don't immediately see why the seed gets passed around separately instead of getting passed through as an internal field of PartBuilder?

It's all about re-creating a `Part` after giving away ownership of the inner columns to encoders. `PartBuilderCfg` (previously `PartBuilderSeed`) is needed because we use `K::Schema` and `V::Schema` to create a `Part`, but then we want to take ownership of the columns within the `Part` so we can create encoders. But we also need to retain the `DynStructCfg`s from the `Part` so we can later put everything back together.

Two alternatives are:
1. We change `Schema::encoder(cols: ColumnsMut)` to be `Schema::encoder(cfg: DynStructCfg, cols: ColumnsMut)` but the encoder doesn't need the `DynStructCfg` for anything, so I didn't love this approach.
2. We add a new method like `Part::from_cols(key_schema, val_schema, key_col, val_col, ts_col, diff_col)`, but it felt bad to creating a `Part` to then call `Part::from_cols(...)` later on.

Typing this up now I realized the best approach might be something like:
1. Add a new method `ColumnsMut::from_schema(&schema)`, use this to create encoders.
2. Add `Part::from_cols(...)` as described above.
This requires us to provide the `schema` twice, once to `ColumnsMut::from_schema` and again to `Part::from_cols(...)` which seems redundant, but is maybe not too bad?

Reply:
Okay, I've talked myself into the necessity of exploding PartBuilder out to components and collecting it back together. The abstraction leakage is _really_ unfortunate, but I do think this is better than trying to do anything with self-referencing types.

The next bit I want to zoom in on is whether we can remove _some_ of the new boilerplate (and abstraction leakage). In something like this, can we move the `.encoder` and `.finish` calls into `PartBuilder::new` and `into_part` respectively by adding some sort of `<K, V>` bounds to `PartBuilder`?

```
    let (cfg, builder) = PartBuilder::new(schema, &UNIT_SCHEMA);
    let PartBuilder {
        key,
        val,
        mut ts,
        mut diff,
    } = builder;

    let mut key_encoder = schema.encoder(key)?;
    let mut val_encoder = UNIT_SCHEMA.encoder(val)?;

    key_encoder.encode(value);
    val_encoder.encode(&());
    ts.push(1u64);
    diff.push(1i64);

    let key_columns = key_encoder.finish();
    let val_columns = val_encoder.finish();

    let part = cfg.into_part(key_columns, val_columns, ts, diff)?;
```

Reply:
Dan and I chatted offline, we came up with a much better API for `PartBuilder` that prevent the need to manage so many different types and function calls. Updated the PR with this new refactor.

Note, now the API looks like:
```
let mut builder = PartBuilder::new(&key_schema, &val_schema);
for (key, val, ts, diff) in rows {
  builder.push(key, val, ts, diff);
}
let part = builder.finish();
```

This is much cleaner than before, but also makes it feel like we should implement some `.extend(<list>)` function that can pre-allocate space, e.g.
```
builder.extend(updates);
```
I'll add this in a followup PR